### PR TITLE
feat: support deletion actions for foreign keys in table editor

### PR DESCRIPTION
**Pull request by joshenlim**
- Support deletion action for foreign keys in table editor
  - You can now set a deletion action when setting a foreign key
  
  

- Misc UI improvements:
  - Tooltip added in table side panel editor
image
image
  - Tooltip added in grid column for foreign keys
image
  - Added a cancel button if trying to remove foreign key
image

### CODE REVIEW

File: studio/components/interfaces/TableGridEditor/SidePanelEditor/ForeignKeySelector/ForeignKeySelector.utils.tsx [hunk 1]
```diff
@@ -0,0 +1,46 @@
+import { FOREIGN_KEY_DELETION_ACTION } from 'data/database/database-query-constants'
+
+export const generateDeletionActionDescription = (deletionAction: string, reference: string) => {
+  switch (deletionAction) {
+    case FOREIGN_KEY_DELETION_ACTION.NO_ACTION:
+      return (
+        <>
+          Deleting a record from <code className="text-xs text-scale-1100">{reference}</code> will{' '}
+          <span className="text-brand-900 opacity-75">raise an error</span> if there exists
+          referencing rows from this table
```
Comment (alaister) [inline studio/components/interfaces/TableGridEditor/SidePanelEditor/ForeignKeySelector/ForeignKeySelector.utils.tsx:10]: "if there exists referencing rows from this table" -> "if there are records existing in this table that reference it"

File: studio/components/interfaces/TableGridEditor/SidePanelEditor/ForeignKeySelector/ForeignKeySelector.utils.tsx [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+        </>
+      )
+    case FOREIGN_KEY_DELETION_ACTION.CASCADE:
+      return (
+        <>
+          Deleting a record from <code className="text-xs text-scale-1100">{reference}</code> will{' '}
+          <span className="text-brand-900 opacity-75">also delete</span> the referencing records
+          from this table
```
Comment (alaister) [inline studio/components/interfaces/TableGridEditor/SidePanelEditor/ForeignKeySelector/ForeignKeySelector.utils.tsx:18]: "the referencing records from this table" -> "any records that reference it in this table"

File: studio/components/interfaces/TableGridEditor/SidePanelEditor/ForeignKeySelector/ForeignKeySelector.utils.tsx [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+        </>
+      )
+    case FOREIGN_KEY_DELETION_ACTION.RESTRICT:
+      return (
+        <>
+          Deleting a record from <code className="text-xs text-scale-1100">{reference}</code> will{' '}
+          <span className="text-brand-900 opacity-75">prevent deletion</span> of existing
+          referencing rows from this table
```
Comment (alaister) [inline studio/components/interfaces/TableGridEditor/SidePanelEditor/ForeignKeySelector/ForeignKeySelector.utils.tsx:26]: This is essentially the same as "no action", except it's not deferrable, which is something I don't think most users will care about. What do you think we should be saying here to reflect this?

https://www.postgresql.org/docs/current/sql-createtable.html

File: studio/components/interfaces/TableGridEditor/SidePanelEditor/ForeignKeySelector/ForeignKeySelector.utils.tsx [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+        </>
+      )
+    case FOREIGN_KEY_DELETION_ACTION.SET_DEFAULT:
+      return (
+        <>
+          Deleting a record from <code className="text-xs text-scale-1100">{reference}</code> will{' '}
+          set the values of existing referencing rows to their{' '}
+          <span className="text-brand-900 opacity-75">default value</span> from this table
```
Comment (alaister) [inline studio/components/interfaces/TableGridEditor/SidePanelEditor/ForeignKeySelector/ForeignKeySelector.utils.tsx:34]: "will set the values of existing referencing rows to their default value from this table" -> "will set the value of any existing records in this table referencing it back to their default value"

File: studio/components/interfaces/TableGridEditor/SidePanelEditor/ForeignKeySelector/ForeignKeySelector.utils.tsx [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+        </>
+      )
+    case FOREIGN_KEY_DELETION_ACTION.SET_NULL:
+      return (
+        <>
+          Deleting a record from <code className="text-xs text-scale-1100">{reference}</code> will{' '}
+          set the values of existing referencing rows{' '}
+          <span className="text-brand-900 opacity-75">to NULL</span> from this table
```
Comment (alaister) [inline studio/components/interfaces/TableGridEditor/SidePanelEditor/ForeignKeySelector/ForeignKeySelector.utils.tsx:42]: "will set the values of existing referencing rows to NULL from this table" -> "will set the value of any existing records in this table referencing it to NULL"

File: studio/data/database/foreign-key-constraints-query.ts [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+    (foreignKey: ForeignKeyConstraint) => {
+      return {
+        ...foreignKey,
+        source_columns: foreignKey.source_columns.replace('{', '').replace('}', '').split(','),
+        target_columns: foreignKey.target_columns.replace('{', '').replace('}', '').split(','),
+      }
+    }
+  )
+
+  return { ...response, data: { ...response.data, result: formattedResults } }
```
Comment (alaister) [inline studio/data/database/foreign-key-constraints-query.ts:106]: You might be able to use the `select` option in react-query to do this transformation instead of outside. It should make the types easier for you.

https://tanstack.com/query/v4/docs/react/reference/useQuery#:~:text=select%3A%20(data%3A%20TData,the%20query%20cache.

**Review by alaister — CHANGES_REQUESTED** [review]
Working great! 🔥

I've left some feedback regarding the wording of the helper text for choosing the deletion action. I find this to be the most confusing part of foreign keys!

I found the word 'referencing' to be very dense to parse in my brain box, so I've added some suggestions that swap it for 'references it', however, I'm not sure if they're better. They're more verbose and it's easy to lose track of what 'it' is. I think we should **get a third party to weigh in here**!

File: studio/components/interfaces/TableGridEditor/SidePanelEditor/ForeignKeySelector/ForeignKeySelector.tsx [hunk 7]
```diff
@@ -240,6 +272,25 @@ const ForeignKeySelector: FC<Props> = ({ column, visible = false, closePanel, sa
                   ))}
                 </Listbox>
               )}
+              <SidePanel.Separator />
+              <Listbox
+                id="deletionAction"
+                value={selectedForeignKey.deletionAction}
+                label="Deletion action if referenced row is removed"
```
Comment (saltcod) [inline studio/components/interfaces/TableGridEditor/SidePanelEditor/ForeignKeySelector/ForeignKeySelector.tsx:279]: I think this should be "Delete action if...."

File: studio/components/interfaces/TableGridEditor/SidePanelEditor/ForeignKeySelector/ForeignKeySelector.utils.tsx (hunk 1 above)
Comment (saltcod) [reply studio/components/interfaces/TableGridEditor/SidePanelEditor/ForeignKeySelector/ForeignKeySelector.utils.tsx:10]: Also think think the text should be underlined here instead of green, as it looks a bit like a link.

Comment (saltcod) [reply studio/components/interfaces/TableGridEditor/SidePanelEditor/ForeignKeySelector/ForeignKeySelector.utils.tsx:10]: hummm it wasn't super clear to me that this text below was contextual. The dropdown covered the text, so I didn't realize it changed based on the dropdown. I did realize eventually ....but maybe worth a bigger callout, and repeat the heading?

CleanShot 2023-03-07 at 12 33 23@2x

File: studio/components/interfaces/TableGridEditor/SidePanelEditor/ForeignKeySelector/ForeignKeySelector.utils.tsx (hunk 3 above)
Comment (saltcod) [reply studio/components/interfaces/TableGridEditor/SidePanelEditor/ForeignKeySelector/ForeignKeySelector.utils.tsx:26]: I think we should put in a little bit of guidance text at the bottom here to help with the confidence factor. What if I don't know which one I want? What if I don't care at this time? What do most people choose? Can I change it later? Can we add a small section to a docs page to cover this in more detail?

CleanShot 2023-03-07 at 12 39 09@2x

Comment (joshenlim) [reply studio/components/interfaces/TableGridEditor/SidePanelEditor/ForeignKeySelector/ForeignKeySelector.utils.tsx:26]: agreed - i was pretty stuck on the explanation here tbh. cause staging "not deferrable" isn't clear as well for devs new to SQL, lemme think about this a little more

File: studio/components/interfaces/TableGridEditor/SidePanelEditor/ForeignKeySelector/ForeignKeySelector.utils.tsx (hunk 1 above)
Comment (joshenlim) [reply studio/components/interfaces/TableGridEditor/SidePanelEditor/ForeignKeySelector/ForeignKeySelector.utils.tsx:10]: hmm i'm not inclined to render a call out here as it takes up quite a bit of real-estate + attention
in the future we'll also support upgrade action for foreign keys as well which will be a similar UI to this

but i do agree that the description is not that visible cause the dropdown is hiding it so the change in desc is not apparent

Comment (joshenlim) [reply studio/components/interfaces/TableGridEditor/SidePanelEditor/ForeignKeySelector/ForeignKeySelector.utils.tsx:10]: im thinking we stick to using the description text here but 
- show the deletion action again as suggested
- highlight the consequence with amber instead of green

image

File: studio/components/interfaces/TableGridEditor/SidePanelEditor/ForeignKeySelector/ForeignKeySelector.tsx (hunk 7 above)
Comment (joshenlim) [reply studio/components/interfaces/TableGridEditor/SidePanelEditor/ForeignKeySelector/ForeignKeySelector.tsx:279]: I realised "Delete action if referenced row is removed" kinda sounds weird as well now that I think twice about it
- kinda implicitly says that my data might get deleted if the reference row is removed, but the selection here is more of a "response" if the reference row is removed

Comment (joshenlim) [reply studio/components/interfaces/TableGridEditor/SidePanelEditor/ForeignKeySelector/ForeignKeySelector.tsx:279]: Maybe just "Action if reference row is removed for now"

File: studio/components/interfaces/TableGridEditor/SidePanelEditor/ForeignKeySelector/ForeignKeySelector.utils.tsx (hunk 3 above)
Comment (joshenlim) [reply studio/components/interfaces/TableGridEditor/SidePanelEditor/ForeignKeySelector/ForeignKeySelector.utils.tsx:26]: great shout on the info box for deciding which action to select, updated as such with an expandable info box:

image

image

Comment (joshenlim) [reply studio/components/interfaces/TableGridEditor/SidePanelEditor/ForeignKeySelector/ForeignKeySelector.utils.tsx:26]: no worries about the docs part - these are postgres specific information, so i'm linking to the docs from postgres here
the info in this infobox is a paraphrasing of what they have on their docs

Comment (joshenlim) [reply studio/components/interfaces/TableGridEditor/SidePanelEditor/ForeignKeySelector/ForeignKeySelector.utils.tsx:26]: opting to show a tooltip to explain the difference between restrict and no action, trying not to have even more information up front on the UI than we already have - it's getting a bit too wordy and the UX may be degraded

image

File: studio/components/interfaces/TableGridEditor/SidePanelEditor/ForeignKeySelector/ForeignKeySelector.tsx [hunk 8]
```diff
@@ -240,6 +272,60 @@ const ForeignKeySelector: FC<Props> = ({ column, visible = false, closePanel, sa
...
+                      </li>
+                      <li>
+                        <code className="text-xs">Set NULL</code> or{' '}
+                        <code className="text-xs">Set default</code>: if a foreign-key relationship
+                        represents optional information
+                      </li>
+                    </ul>
+                    <p className="mt-2">
+                      Typically, restricting and cascading deletes are the most common options, but
+                      the default behaviour is no action
```
[the remark this answers is not in the document]
Comment (joshenlim) [reply studio/components/interfaces/TableGridEditor/SidePanelEditor/ForeignKeySelector/ForeignKeySelector.tsx:304]: okay fine

**Comment by joshenlim** [comment]
thanks so much for all the feedback here guys - really appreciate them, super helpful! 🙏

**Review by alaister — APPROVED** [review]
LGTM!

**Comment by AdamEisfeld** [comment]
Any idea on if/when this will come to the localhost studio editor as well?
